// dnd-core/modules/Flux.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

exports.__esModule = true;

var _Flummox2 = require('flummox');

var _DragDropActions = require('./actions/DragDropActions');

var _DragDropActions2 = _interopRequireWildcard(_DragDropActions);

var _RegistryActions = require('./actions/RegistryActions');

var _RegistryActions2 = _interopRequireWildcard(_RegistryActions);

var _DragOperationStore = require('./stores/DragOperationStore');

var _DragOperationStore2 = _interopRequireWildcard(_DragOperationStore);

var _DragOffsetStore = require('./stores/DragOffsetStore');

var _DragOffsetStore2 = _interopRequireWildcard(_DragOffsetStore);

var _RefCountStore = require('./stores/RefCountStore');

var _RefCountStore2 = _interopRequireWildcard(_RefCountStore);

var Flux = (function (_Flummox) {
  function Flux(manager) {
    _classCallCheck(this, Flux);

    _Flummox.call(this);

    this.dragDropActions = this.createActions('dragDropActions', _DragDropActions2['default'], manager);
    this.dragDropActionIds = this.getActionIds('dragDropActions');

    this.registryActions = this.createActions('registryActions', _RegistryActions2['default']);
    this.registryActionIds = this.getActionIds('registryActions');

    this.dragOperationStore = this.createStore('dragOperationStore', _DragOperationStore2['default'], this);

    this.dragOffsetStore = this.createStore('dragOffsetStore', _DragOffsetStore2['default'], this);

    this.refCountStore = this.createStore('refCountStore', _RefCountStore2['default'], this);
  }

  _inherits(Flux, _Flummox);

  return Flux;
})(_Flummox2.Flummox);

exports['default'] = Flux;
module.exports = exports['default'];
