// dnd-core/modules/actions/RegistryActions.js
'use strict';

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

exports.__esModule = true;

var _Actions2 = require('flummox');

var RegistryActions = (function (_Actions) {
  function RegistryActions() {
    _classCallCheck(this, RegistryActions);

    if (_Actions != null) {
      _Actions.apply(this, arguments);
    }
  }

  _inherits(RegistryActions, _Actions);

  RegistryActions.prototype.addSource = function addSource(sourceId) {
    return { sourceId: sourceId };
  };

  RegistryActions.prototype.addTarget = function addTarget(targetId) {
    return { targetId: targetId };
  };

  RegistryActions.prototype.removeSource = function removeSource(sourceId) {
    return { sourceId: sourceId };
  };

  RegistryActions.prototype.removeTarget = function removeTarget(targetId) {
    return { targetId: targetId };
  };

  return RegistryActions;
})(_Actions2.Actions);

exports['default'] = RegistryActions;
module.exports = exports['default'];
