// flummox/lib/addons/FluxComponent.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

/**
 * Flux Component
 *
 * Component form of fluxMixin. Uses fluxMixin as part of its implementation,
 * so requires a flux instance to be provided as either context or a prop.
 *
 * Like fluxMixin, children are given access to the flux instance via
 * `context.flux`. Use this near the top of your app hierarchy and all children
 * will have easy access to the flux instance (including, of course, other
 * Flux components!):
 *
 * <FluxComponent flux={flux}>
 *    ...the rest of your app
 * </FluxComponent>
 *
 * Now any child can access the flux instance again like this:
 *
 * <FluxComponent>
 *    ...children
 * </FluxComponent>
 *
 * We don't need the flux prop this time because flux is already part of
 * the context.
 *
 * Additionally, immediate children are given a `flux` prop.
 *
 * The component has an optional prop `connectToStores`, which is -- you guessed
 * it -- passed directly to fluxMixin's `connectToStores()` function and
 * set as the initial state. The component's state is injected as props to
 * child components.
 *
 * The practical upshot of all this is that fluxMixin, state changes, and
 * context are now simply implementation details. Among other things, this means
 * you can write your components as plain ES6 classes. Here's an example:
 *
 * class ParentComponent extends React.Component {
 *
 *   render() {
 *     <FluxComponent connectToStore="fooStore">
 *       <ChildComponent />
 *     </FluxComponent>
 *   }
 *
 * }
 *
 * ChildComponent in this example has prop `flux` containing the flux instance,
 * and props that sync with each of the state keys of fooStore.
 */

var React = _interopRequire(require("react/addons"));

var fluxMixin = _interopRequire(require("./fluxMixin"));

var assign = _interopRequire(require("object-assign"));

var FluxComponent = React.createClass({
  displayName: "FluxComponent",


  mixins: [fluxMixin()],

  getInitialState: function getInitialState() {
    return this.connectToStores(this.props.connectToStores);
  },

  wrapChild: function wrapChild(child) {
    var _props = this.props;
    var children = _props.children;
    var connectToStores = _props.connectToStores;
    var props = _objectWithoutProperties(_props, ["children", "connectToStores"]);

    return React.addons.cloneWithProps(child, assign({
      flux: this.flux }, this.state, props));
  },

  render: function render() {
    var children = this.props.children;


    if (!children) {
      return null;
    }if (!Array.isArray(children)) {
      var child = children;
      return this.wrapChild(child);
    } else {
      return React.createElement(
        "span",
        null,
        React.Children.map(children, this.wrapChild)
      );
    }
  }

});

module.exports = FluxComponent;
